pkg/kubelet/metrics/collectors/cri_metrics.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collectors
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/component-base/metrics"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+)
+
+type criMetricsCollector struct {
+	metrics.BaseStableCollector
+	listPodSandboxMetrics func(context.Context) ([]*runtimeapi.PodSandboxMetrics, error)
+}
+
+// Check if criMetricsCollector implements necessary interface
+var _ prometheus.Collector = &criMetricsCollector{}
+
+// NewCRIMetricsCollector implements the prometheus.Collector interface
+func NewCRIMetricsCollector(listPodSandboxMetrics func(context.Context) ([]*runtimeapi.PodSandboxMetrics, error)) prometheus.Collector {
+	return &criMetricsCollector{
+		listPodSandboxMetrics: listPodSandboxMetrics,
+	}
+}
+
+// Describe implements the prometheus.Describe interface.
+// TODO(haircommander): would it be better if these were processed async?
+func (c *criMetricsCollector) Describe(ch chan<- *prometheus.Desc) {
```

Comment:
Do you know when Describe and when Collect are collect? If for every scrape describe gets called we end up having to make two calls to the CRI :( 

Maybe we should be following the pattern of making the descriptors on the fly similar to cAdvisor approach? - https://github.com/google/cadvisor/blob/master/metrics/prometheus.go#L1900

Though we can always optimize this later, happy to start with this approach and iterate.

Reply:
we're currently creating them on the fly, we'd have to cache them if we wanted to save the value (which I'm open to, but would further complicate kubelet implementation and cause drift)

Reply:
Describe is called once at the startup, IIRC.  You are allowed to not implement Describe, but it removes a lot of important validation.